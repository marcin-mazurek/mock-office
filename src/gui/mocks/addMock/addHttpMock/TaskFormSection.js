import React from 'react';
import Select from 'react-select';
import { Field } from 'redux-form/immutable';

const TaskType = field =>
  <Select
    name="server-type"
    value={field.input.value}
    onChange={option => field.input.onChange(option.value)}
    searchable={false}
    clearable={false}
    className="form-field__select"
    options={[
      { value: 'immediate', label: 'immediate' },
      { value: 'future', label: 'future' }
    ]}
  />
;

const TaskFormSection = () => (
  <div style={{ width: '100%' }}>
    <div className="form-row">
      <div className="form__field">
        <label className="form-field__label" htmlFor="title">Title:</label>
        <Field
          className="form-field__input"
          component="input"
          name="title"
          type="text"
          placeholder="task title"
        />
      </div>
    </div>
    <div className="form-row">
      <div className="form__field">
        <label className="form-field__label" htmlFor="type">Type:</label>
        <Field component={TaskType} name="type" />
      </div>
    </div>
    <div className="form-row">
      <div className="form__field">
        <label className="form-field__label" htmlFor="payload">Payload:</label>
        <Field component="textarea" name="payload" />
      </div>
    </div>
    <div className="form-row">
      <div className="form__field">
        <label className="form-field__label" htmlFor="delay">Delay:</label>
        <Field
          className="form-field__input"
          component="input"
          type="number"
          name="delay"
        />
      </div>
    </div>
  </div>
);

export default TaskFormSection;
