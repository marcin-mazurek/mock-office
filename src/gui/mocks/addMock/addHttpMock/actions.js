export const SUBMIT = 'addHttpMock/SUBMIT';

export const submit = (server, scenario, formValues) => ({
  type: SUBMIT,
  server,
  scenario,
  formValues
});
