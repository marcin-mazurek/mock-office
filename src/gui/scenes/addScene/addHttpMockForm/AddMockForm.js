import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm, FormSection } from 'redux-form/immutable';
import { connect } from 'react-redux';
import ScenePartFormSection from '../ScenePartFormSection';
import { submitHttpScene } from './addSceneEpic';

export const AddMockForm = props => (
  <form className="form" onSubmit={props.handleSubmit}>
    <div className="form__header">
      Add mock
    </div>
    <div className="form-row">
      <div className="form__field">
        <label className="form-field__label" htmlFor="title">Title:</label>
        <Field
          name="title"
          component="input"
          type="text"
          placeholder="Mock title"
          className="form-field__input"
        />
      </div>
    </div>
    <section className="form-section">
      <header className="form-section__header">Request:</header>
      <div className="form-row">
        <div className="form__field">
          <label className="form-field__label" htmlFor="requirements">Requirements:</label>
          <Field
            name="requirements"
            component="textarea"
            cols="30"
            rows="10"
          />
        </div>
      </div>
    </section>
    <section className="form-section">
      <header className="form-section__header">Response:</header>
      <FormSection name="scenePart">
        <ScenePartFormSection />
      </FormSection>
      <div className="form-row">
        <button className="button form__button" type="submit">
          Submit
        </button>
      </div>
    </section>
  </form>
);

AddMockForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  onSubmit: values => dispatch(submitHttpScene(ownProps.scenarioId, values.toJS()))
});

export default connect(null, mapDispatchToProps)(
  reduxForm(
    {
      form: 'httpMock',
      initialValues: {
        scenePart: {
          type: 'immediate'
        }
      }
    }
  )(AddMockForm)
);
