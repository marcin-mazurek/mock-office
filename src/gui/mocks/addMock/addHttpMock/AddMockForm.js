import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm, FormSection } from 'redux-form/immutable';
import { connect } from 'react-redux';
import TaskFormSection from './TaskFormSection';
import { submit } from './actions';

export const AddMockForm = props => (
  <form className="form" onSubmit={props.handleSubmit}>
    <div className="form__header">
      Add mock
    </div>
    <div className="form-row">
      <div className="form__field">
        <label className="form-field__label" htmlFor="title">Title:</label>
        <Field
          name="title"
          component="input"
          type="text"
          placeholder="Mock title"
          className="form-field__input"
        />
      </div>
    </div>
    <section className="form-section">
      <header className="form-section__header">Request:</header>
      <div className="form-row">
        <div className="form__field">
          <label className="form-field__label" htmlFor="requirements">Requirements:</label>
          <Field
            name="requirements"
            component="textarea"
            cols="30"
            rows="10"
          />
        </div>
      </div>
    </section>
    <section className="form-section">
      <header className="form-section__header">Response:</header>
      <FormSection name="task">
        <TaskFormSection />
      </FormSection>
      <div className="form-row">
        <button className="button form__button" type="submit">
          Submit
        </button>
      </div>
    </section>
  </form>
);

AddMockForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  onSubmit: values => dispatch(submit(ownProps.server, ownProps.scenario, values.toJS()))
});

const AddMockFormConnect = connect(null, mapDispatchToProps)(
  reduxForm(
    {
      form: 'addHttpMock',
      initialValues: {
        task: {
          type: 'immediate'
        }
      }
    }
  )(AddMockForm)
);

export default AddMockFormConnect;
