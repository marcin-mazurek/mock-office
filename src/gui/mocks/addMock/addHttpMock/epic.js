import { Observable } from 'rxjs';
import { ifElse, has } from 'ramda';
import api from '../../../resources/api';
import { addAction as addNotification } from '../../../notifications/actions';
import { SUBMIT } from './actions';

export const SUCCEED = 'addMock/SUCCEED';
export const succeedAction = (scenario, mock) => ({
  type: SUCCEED,
  scenario,
  mock
});
export const FAILED = 'addMock/FAILED';
export const failedAction = reason => ({
  type: FAILED,
  reason
});

const hasError = has('error');
const processFormValues = (formValues) => {
  const fV = formValues;
  let requirements;

  const requirementsSubmitted = fV.requirements;
  if (requirementsSubmitted) {
    try {
      requirements = JSON.parse(requirementsSubmitted);
    } catch (error) {
      throw new Error('Requirements JSON is broken');
    }
    requirements = Object.assign({}, requirements);
  } else {
    requirements = {};
  }

  if (fV.task && fV.task.payload) {
    try {
      fV.task.payload = JSON.parse(fV.task.payload);
    } catch (error) {
      throw new Error('Payload JSON is broken');
    }
  }

  if (fV.task.delay) {
    if (fV.task.delay < 0) {
      return {
        error: {
          message: 'Delay is < 0'
        }
      };
    }

    fV.task.delay = parseInt(fV.task.delay, 10);
  }

  return {
    data: {
      title: fV.title,
      requirements,
      tasks: [fV.task]
    }
  };
};

const onFail = result => failedAction(result.error);
const onSuccess = result => succeedAction(result.data.scenario, result.data.mock);

export default function addMockEpic(action$) {
  return action$.ofType(SUBMIT)
    .flatMap((action) => {
      try {
        const { data, error } = processFormValues(action.formValues);

        if (error) {
          return Promise.resolve({ error });
        }

        const reqWithEvent = Object.assign({}, data.requirements, { event: 'RECEIVED_REQUEST' });

        return Observable.from(
          api.addMock(action.server, action.scenario, Object.assign(data,
            {
              requirements: reqWithEvent
            }
          )));
      } catch (error) {
        return Observable.of(addNotification({ text: error.message, type: 'error' }));
      }
    })
    .map(
      ifElse(
        hasError,
        onFail,
        onSuccess
      )
    );
}
