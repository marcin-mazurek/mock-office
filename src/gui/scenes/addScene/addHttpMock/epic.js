import { Observable } from 'rxjs';
import { requestAddScene } from '../../../api/api';
import { add as addNotification } from '../../../entities/notifications/actions';
import { add } from '../actions';

export const SUBMIT = 'addMock/SUBMIT';

export const submit = (scenarioId, formValues) => ({
  type: SUBMIT,
  scenarioId,
  formValues
});

const processFormValues = (formValues) => {
  const fV = formValues;
  let requirements;

  try {
    const requirementsSubmitted = fV.requirements;
    if (requirementsSubmitted) {
      requirements = JSON.parse(requirementsSubmitted);
      requirements = Object.assign({}, requirements);
    } else {
      requirements = {};
    }

    if (fV.scenePart && fV.scenePart.payload) {
      fV.scenePart.payload = JSON.parse(fV.scenePart.payload);
    }
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error.message);
  }

  if (fV.scenePart.delay) {
    fV.scenePart.delay = parseInt(fV.scenePart.delay, 10);
  }

  return {
    title: fV.title,
    requirements,
    parts: [fV.scenePart]
  };
};

export default function addMockEpic(action$) {
  return action$.ofType(SUBMIT)
    .flatMap((action) => {
      const sceneParams = processFormValues(action.formValues);

      return Observable.from(
        Promise.all(
          [Object.assign(sceneParams, { event: 'RECEIVED_REQUEST' })].map(
            scene => requestAddScene(action.scenarioId, scene)
          )
        )
      );
    })
    .flatMap(scenes => Observable.from(scenes))
    .map(scene => add(...scene))
    .catch(err => Observable.of(addNotification({ text: err.message, type: 'error' })));
}
