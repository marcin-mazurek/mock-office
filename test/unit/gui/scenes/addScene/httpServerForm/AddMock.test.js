import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AddMockForm } from '../../../../../../src/gui/mocks/addMock/addHttpMock/AddMockForm';

describe('HttpForm', () => {
  test('default snapshot', () => {
    const props = {
      queueId: 'some id',
      handleSubmit: () => {}
    };
    const wrapper = shallow(<AddMockForm {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
