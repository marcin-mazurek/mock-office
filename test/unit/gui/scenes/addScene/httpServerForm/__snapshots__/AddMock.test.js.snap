// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpForm default snapshot 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="form-row"
  >
    <div
      className="form__field"
    >
      <label
        className="form-field__label"
        htmlFor="title"
      >
        Title:
      </label>
      <Field
        className="form-field__input"
        component="input"
        name="title"
        placeholder="Mock title"
        type="text"
      />
    </div>
  </div>
  <section
    className="form-section"
  >
    <header
      className="form-section__header"
    >
      Request:
    </header>
    <div
      className="form-row"
    >
      <div
        className="form__field"
      >
        <label
          className="form-field__label"
          htmlFor="requirements"
        >
          Requirements:
        </label>
        <Field
          cols="30"
          component="textarea"
          name="requirements"
          rows="10"
        />
      </div>
    </div>
  </section>
  <section
    className="form-section"
  >
    <header
      className="form-section__header"
    >
      Response:
    </header>
    <FormSection
      component="div"
      name="task"
    >
      <TaskFormSection />
    </FormSection>
    <div
      className="form-row"
    >
      <button
        className="button form__button"
        type="submit"
      >
        Submit
      </button>
    </div>
  </section>
</form>
`;
